pkg/snap/session.go

```diff
@@ -124,114 +117,254 @@ func (s *Session) Start(tags map[string]interface{}) error {
+	if err != nil {
+		return err
+	}
+
+	if coreStatus == core.TaskDisabled {
+		return errors.Errorf("snap task %q has been disabled because of errors: %q", s.task.Name, s.lastFailureMessage)
+	}
+
+	if executorStatus == executor.TERMINATED {
+		return nil
 	}
 
 	rs := s.pClient.StopTask(s.task.ID)
 	if rs.Err != nil {
-		return errors.Wrapf(rs.Err, "could not send stop signal to task %q", s.task.ID)
+		return errors.Wrapf(rs.Err, "could not stop snap task %q: %v", s.task.ID)
 	}
 
-	err := s.waitForStop()
+	err = s.waitForStop()
 	if err != nil {
-		return errors.Wrapf(err, "could not stop task %q", s.task.ID)
+		return errors.Wrapf(err, "could not stop snap task %q", s.task.ID)
 	}
 
-	rr := s.pClient.RemoveTask(s.task.ID)
-	if rr.Err != nil {
-		return errors.Wrapf(rr.Err, "could not remove task %q", s.task.ID)
+	return nil
+}
+
+// Wait blocks until the taskInfo is executed at least once
+// (including hits that happened in the past).
+func (s *Handle) Wait(timeout time.Duration) (bool, error) {
+	executorStatus, coreStatus, err := s.getStatus()
+	if err != nil {
+		return true, err
 	}
 
-	s.task = nil
+	if coreStatus == core.TaskDisabled {
+		return true, errors.Errorf("snap task %q has been disabled because of errors: %q", s.task.Name, s.lastFailureMessage)
+	}
 
-	return nil
+	if executorStatus == executor.TERMINATED {
+		return true, nil
+	}
+
+	// Task has not finished yet.
+
+	stopper := make(chan struct{})
+	taskCompletionInfo := s.waitForTaskCompletion(stopper)
+
+	var timeoutChannel <-chan time.Time
+	if timeout != 0 {
+		// In case of wait with timeout set the timeout channel.
+		timeoutChannel = time.After(timeout)
+	}
+
+	select {
+	case err := <-taskCompletionInfo:
+		// If waitEndChannel is closed then taskInfo is terminated.
+		return true, err
+	case <-timeoutChannel:
+		// If timeout time exceeded return then taskInfo did not terminate yet.
+		stopper <- struct{}{}
+		return false, nil
+	}
 }
 
-// Wait blocks until the task is executed at least once
-// (including hits that happened in the past).
-func (s *Session) Wait() error {
-	for {
-		t := s.pClient.GetTask(s.task.ID)
-		if t.Err != nil {
-			return errors.Wrapf(t.Err, "getting task %q failed", s.task.ID)
-		}
+// StdoutFile returns error for snap handles.
+func (s *Handle) StdoutFile() (*os.File, error) {
+	return nil, errors.New("snap tasks don't support stdout file")
+}
 
-		if t.State == "Stopped" || t.State == "Disabled" {
-			return errors.Errorf("failed to wait for task: task %q is in state %q (last failure: %q)",
-				s.task.ID,
-				t.State,
-				t.LastFailureMessage)
+// StderrFile returns error for snap handles.
+func (s *Handle) StderrFile() (*os.File, error) {
+	return nil, errors.New("snap tasks don't support stderr file")
+}
 
-		}
+// EraseOutput does nothing for snap tasks.
+func (s *Handle) EraseOutput() error {
+	return nil
+}
+
+func (s *Handle) waitForTaskCompletion(stopper <-chan struct{}) <-chan error {
```

Comment:
@skonefal - I think it would be more consistent if we choose one of options:
* pass both channels as arguments to the function;
* return both channels from the function.